File: next/dweb-browser/Helper/src/Changeable.cs
namespace DwebBrowser.Helper;

public class Changeable<T>
{
    private T Self { get; init; }
    public Listener<T> Listener = Listener<T>.New();

    public Changeable(T self)
    {
        Self = self;
        Watch = new(Listener, self);
    }

    public Watcher Watch { get; init; }

    public record Watcher(Listener<T> Listener, T Arg)
    {
        public Task AddWatch(Signal<T> cb) => Listener.AddListenerAndEmitFirst(Arg, cb);
    }

    public Task EmitChangeAsync() => Listener.Emit(Self);

    public void EmitChange()
    {
        _ = EmitChangeAsync();
    }

    /// <summary>
    /// 这里一定要判断size，因为 CoroutineScope 可能已经被 cancel 了，所以只要外部代码正确 onDispose 移除监听，那么这里就不会引发问题
    /// 否则如果listener的size不为空，导致执行了runBlockingCatching带来异常，就说明代码有生命周期问题
    /// </summary>
    public void EmitChangeBackground(T? changes)
    {
        changes ??= Self;

        if (Listener.Size > 0)
        {
            _ = Listener.Emit(changes);
        }
    }
}

File: next/kmp/helper/src/commonMain/kotlin/org/dweb_browser/helper/Changeable.kt
package org.dweb_browser.helper

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

internal class Changeable<T>(
  private val self: T, var context: CoroutineContext,
) {
  private val scope by lazy { CoroutineScope(context) }
  private val signal = lazy { Signal<T>() }
  private inline fun ifSignal(block: (Signal<T>) -> Unit) {
    if (signal.isInitialized()) {
      block(signal.value)
    }
  }

  val onChange by lazy { signal.value.toListener() }

  /**
   * 这里一定要判断size，因为 CoroutineScope 可能已经被 cancel 了，所以只要外部代码正确 onDispose 移除监听，那么这里就不会引发问题
   * 否则如果listener的size不为空，导致执行了runBlockingCatching带来异常，就说明代码有生命周期问题
   */
  fun emitChangeBackground(changes: T = self) = ifSignal { signal ->
    if (signal.size > 0) scope.launch {
      signal.emit(changes)
    }
  }

  suspend fun emitChange(changes: T = self) = ifSignal { signal ->
    signal.emit(changes)
  }

  fun clear() = ifSignal { signal ->
    signal.clear()
  }
}